src/rust/engine/graph/src/tests.rs

```diff
@@ -348,6 +348,34 @@ async fn uncachable_node_only_runs_once() {
   );
 }
 
+#[tokio::test]
+async fn uncachable_deps_node_only_runs_once_per_session() {
+  let _logger = env_logger::try_init();
+  let graph = Arc::new(Graph::new());
+
+  let context = {
+    let mut uncacheable = HashSet::new();
+    uncacheable.insert(TNode::new(1));
+    TContext::new(graph.clone()).with_uncacheable(uncacheable)
+  };
+
+  assert_eq!(
+    graph.create(TNode::new(2), &context).await,
+    Ok(vec![T(0, 0), T(1, 0), T(2, 0)])
+  );
+
+  let expected_node_runs = vec![TNode::new(2), TNode::new(1), TNode::new(0)];
+
+  assert_eq!(context.runs(), expected_node_runs);
+
+  assert_eq!(
+    graph.create(TNode::new(2), &context).await,
+    Ok(vec![T(0, 0), T(1, 0), T(2, 0)])
+  );
+
+  assert_eq!(context.runs(), expected_node_runs);
```

Comment:
Shouldn't we expect another TNode(1) run when we request the graph to create TNode(2) because it is uncacheable. Maybe the underlying implementation has changed since I touched it, but that was my intuition.

Reply:
No, because the second call to `create` uses the same `context` objects, which represents (in this test framework) requesting a node in the same session. Uncacheable nodes should run exactly once per session, so we do expect there to not be another run of `TNode(1)`